Kb gcbias take2

kbergin opened the pull request:
Changed CollectGcBiasMetrics to work as a SinglePassSam

yfarjoun commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -82,13 +71,6 @@
     @Option(shortName = StandardOptionDefinitions.REFERENCE_SHORT_NAME, doc = "The reference sequence fasta file.")
```
I think you need to remove the reference as it is already a standard argument of command_line program. I'm not sure why it was an explicit field before....perhaps someone else can explain.

kbergin replied:
You're right, I'm not sure why it was before. It gave me an error when I had others that were duplicates but didn't have a problem with that one. I'll remove it.

yfarjoun commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -245,19 +195,17 @@ protected int doWork() {
         // Qualify the title with the library name iff it's for a single sample
-        final List<SAMReadGroupRecord> readGroups = sam.getFileHeader().getReadGroups();
+        final List<SAMReadGroupRecord> readGroups = save_header.getReadGroups();
         if (readGroups.size() == 1) {
```
given that this is your only usage of the header, I'd pull out the information you need (during the initialize) and store it (it's a String) rather than holding on to an old header.

nh13 replied:
Also, use camel case.

kbergin replied:
Good idea!

yfarjoun commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -101,107 +83,75 @@
+    static int lastContig = -1;
+    byte[] gc;
+    byte[] refBases;
+    private SAMFileHeader save_header;
```
naming convention for variables: "camelCase" with first letter lower-case

yfarjoun commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -101,107 +83,75 @@
+    final int[] readsByGc = new int[101];
+    final long[] basesByGc = new long[101];
+    final long[] errorsByGc = new long[101];
+    static int lastContig = -1;
```
there's a constant in SamRecord called NO_ALIGNMENT_REFERENCE_INDEX. Use this instead of defining your own.

yfarjoun commented:
My review is possibly not the official review, but I wanted to take a peek...you should not be checking in the picard.iml & picard.ipr files. So you'll need to revert these and re-commit.

mcovarr commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -101,107 +83,75 @@
+    protected void acceptRead(final SAMRecord rec, final ReferenceSequence ref) {
+        //if read is unaligned then ref is passed in as null
+        if (ref == null) {
+            if (!rec.getReadPairedFlag() || rec.getFirstOfPairFlag()) ++this.totalClusters;
```
This statement is on both sides of the `if (ref == null)` conditional.  Can this be pulled above the `if (ref == null)` and then that `if` branch eliminated?  I'm not 100% sure it can, there's some intervening logic in the `else` branch before this statement appears, so this is really a question.  :smile:

kbergin replied:
I think you're right, it could be moved outside of the if and the else could become 
if(ref!=null)

mcovarr commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -101,107 +83,75 @@
+    final int[] windowsByGc = new int[101];
+    final int[] readsByGc = new int[101];
+    final long[] basesByGc = new long[101];
+    final long[] errorsByGc = new long[101];
```
A symbolic constant for the 101 would be good.

kbergin commented:
Ok thank you! I will fix the things you and others have mentioned and then I'm not sure what the Picard.iml and Picard.ipr files are I don't remember touching them but I will figure out where I went wrong and fix it. 

Thanks!
Kylee 



> On Mar 13, 2015, at 4:32 PM, Yossi Farjoun <notifications@github.com> wrote:
> 
> My review is possibly not the official review, but I wanted to take a peek...you should not be checking in the picard.iml & picard.ipr files. So you'll need to revert these and re-commit.
> 
> —
> Reply to this email directly or view it on GitHub.
>

mcovarr commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -101,107 +83,75 @@
+            //only do the recalculation of gc if current ref is different from last ref
+            if (ref.getContigIndex() != lastContig) {
+                refBases = ref.getBases();
+                StringUtil.toUpperCase(refBases);
```
I think you want
```
refBase = StringUtil.toUpperCase(refBases);
```
Java strings are immutable. `StringUtil.toUppercase` returns the uppercase version of its argument, but that's being discarded here.

yfarjoun replied:
@mcovarr I think that this is a verbatim copy of the original code behavior. I agree that as written it is effectively a noop, but since @kbergin is not fixing bugs here we need the results to agree. It might make sense in a subsequent fix to change this, but right now I think she shouldn't change any functional aspect of the code. (my opinion)

tfenne replied:
It's a little harder to see now that the declaration has moved up the class level, but refBases is a byte[], not a String.  So in this case StringUtil.toUpperCase(byte[]) is being called, which does in fact mutate the byte[] so as not to allocate a second copy.

yfarjoun replied:
evil. :-)

kbergin replied:
Cool! Hi Tim!

geoffjentry replied:
![_35](https://cloud.githubusercontent.com/assets/961771/6652405/cd2a1a4a-ca43-11e4-82d4-f32d4b29156b.JPG)

mcovarr replied:
:open_mouth: 

Hi Tim, thanks for the correction and explanation.  I probably should have realized that `StringUtils.toUpperCase` wasn't operating on a `String` since `String#toUpperCase` could do that job.

mcovarr commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -101,107 +83,75 @@
     private int totalAlignedReads = 0;
+    // Histograms to track the number of windows at each GC, and the number of read starts
+    // at windows of each GC
+    final int[] windowsByGc = new int[101];
```
It looks like most of the new fields could be `private`.

kbergin commented on src/java/picard/analysis/CollectGcBiasMetrics.java:
```diff
@@ -211,7 +161,7 @@ protected int doWork() {
         for (int i = 0; i < windowsByGc.length; ++i) {
```
IntelliJ is warning me that the variable minimumWindowsToCountInSummary is never used despite being calculated. Is that something we need? @yfarjoun

yfarjoun replied:
This is also true in the original code...feel free to leave it despite the warning.

yfarjoun commented:
when you are done responding to the reviewer's comments you should comment so that they will not to take a final look and give you thumbs-up!

kbergin commented:
Oh ok, thanks @yfarjoun! Hey Everyone - I think I fixed everything. Reverted the picard files I accidentally pushed. Updated camelCase and changed that variable to a string to get the sample name. I was a little concerned about that because in the original code it was only pulled if the sample being run was a single sample. But I can't think of an instance where more than one sample would be run through this together. Correct me if I'm wrong about that, please! Made variables private and changed the if else to be just an if statement.

The only thing I didn't change and am unsure about still is the "symbolic constant" that @mcovarr suggested. As you mentioned I didn't write that part of the code so I am not sure why 101 was chosen or what the constant should be set to instead. Anyone have an input here?

yfarjoun commented:
Argh!! sorry. found another untested bug. Need to add unit tests!!!

yfarjoun commented:
oops. wrong PR. don't worry @kbergin all is well!

kbergin commented:
Phew!

mcovarr commented:
Regarding the symbolic constant, `private static final int NUM_HISTOGRAM_BUCKETS = 101;` would work for me.  Ideas for better names are welcome!  I'm definitely not suggesting a different value for this constant, I have no idea why this was chosen either.  :smile:

yfarjoun commented:
To allow for nice round numbers including 0 and 100, you need 101 buckets.

On Sat, Mar 14, 2015 at 8:07 PM, mcovarr <notifications@github.com> wrote:

> Regarding the symbolic constant, private static final int
> NUM_HISTOGRAM_BUCKETS = 101; would work for me. Ideas for better names
> are welcome! I'm definitely not suggesting a different value for this
> constant, I have no idea why this was chosen either. [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/168#issuecomment-80768208>.
>

mcovarr commented:
That would be a great Javadoc comment above the constant!  :smile:

kbergin commented:
Hi All - Changed the 101's to constants and pushed again.

kbergin commented:
@mcovarr if it's all good, can I get a thumbs up on this so I can merge and make a new branch to work on something else? :)

mcovarr commented:
Purely a nitpick, but the convention for constants in Java is `ALL_CAPS`, otherwise :+1:

kbergin commented:
@mcovarr Ah thanks! Sorry about that, I'm learning Java as I go. Fixed it!

kbergin commented:
@mcovarr What's the next step to have this merged? Is that something you do?

kbergin commented:
Oh no I did something!! What happened?

kbergin commented:
Phew

mattsooknah commented:
All that's left is to squash it into one commit + rebase on top of master. Then we can merge it.

kbergin commented:
Ok I think I did that @mattsooknah

kbergin commented:
YAY!

yfarjoun commented:
:-)

On Mon, Mar 23, 2015 at 2:19 PM, kbergin <notifications@github.com> wrote:

> YAY!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/168#issuecomment-85131574>.
>
